### PR TITLE
Add remote host validation API 

### PR DESCRIPTION
**Pull request by ChristophWurst**
- [x] Requires https://github.com/nextcloud/server/pull/34847

~~This patch takes the existing internal API ``\OC\Http\Client\LocalAddressChecker`` and makes it more generic and public. This will allow app devs to validate any remote host against Nextcloud's restrictions.~~

This patch generalizes the old "local address checker" into a validator for any remote hosts so that apps can use it.

* Add hostname classifier
* Add ip address classifier
* Add new injectable service (interface+impl)
* Write unit tests for all components
* Add integration test to ensure the parts still work when combined

Documentation can be found at https://github.com/nextcloud/documentation/pull/9259.

### CODE REVIEW

File: lib/private/Http/Client/Client.php [hunk 1]
```diff
@@ -181,7 +184,11 @@ protected function preventLocalAddress(string $uri, array $options): void {
 			return;
 		}
 
-		$this->localAddressChecker->ThrowIfLocalAddress($uri);
+		$host = parse_url($uri, PHP_URL_HOST);
+		if ($host === false || $host === null) {
+			throw new LocalServerException('Could not detect any host');
+		}
+		$this->localAddressChecker->throwIfLocalAddress(urldecode($host));
```
Comment (nickvergessen) [inline lib/private/Http/Client/Client.php:191]: `throwIfLocalAddress` still urldecodes?
Should be only one place?

Comment (ChristophWurst) [reply lib/private/Http/Client/Client.php]: Right. Moved it back and forth. But doubly decoded is not what we want :)

**Review by nickvergessen — APPROVED** [review]

**Comment by nickvergessen** [comment]
Base PR merged

**Review by miaulalala — APPROVED** [review]

**Comment by nickvergessen** [comment]
> ## There were 27 failures:
>
> 1) Test\Http\Client\LocalAddressCheckerTest::testThrowIfLocalAddress with data set #0 ('localhost/foo.bar')
> Failed asserting that exception of type "OCP\Http\Client\LocalServerException" is thrown.
>
> ...

**Comment by ChristophWurst** [comment]
Let me move the old validator into the new one right away

**Comment by ChristophWurst** [comment]
CI is dark-green because of the usual S3 and sharing integration test failure. The rest looks fine.

File: lib/private/Server.php [hunk 2]
```diff
@@ -858,7 +856,7 @@ public function __construct($webRoot, \OC\Config $config) {
 		$this->registerAlias(\OCP\Security\ISecureRandom::class, SecureRandom::class);
 		/** @deprecated 19.0.0 */
 		$this->registerDeprecatedAlias('SecureRandom', \OCP\Security\ISecureRandom::class);
-
+		$this->registerAlias(\OCP\Security\IRemoteHostValidator::class, \OC\Security\RemoteHostValidator::class);
```
Comment (miaulalala) [inline lib/private/Server.php:859]: Import statement?

**Review by miaulalala — APPROVED** [review]

**Review by come-nc — APPROVED** [review]

**Comment by miaulalala** [comment]
Drone failure unrelated

File: lib/private/Http/Client/DnsPinMiddleware.php [hunk 3]
```diff
@@ -133,7 +134,10 @@ public function addDnsPinning() {
 					$curlResolves["$hostName:$port"] = [];
 
 					foreach ($targetIps as $ip) {
-						$this->localAddressChecker->throwIfLocalIp($ip);
+						if (!$this->ipAddressClassifier->isLocalAddress($ip)) {
```
Comment (ChristophWurst) [inline lib/private/Http/Client/DnsPinMiddleware.php:137]: `!` shouldn't be there

Comment (nickvergessen) [reply lib/private/Http/Client/DnsPinMiddleware.php:137]: ops 😆

**Pull request merged**